lib/fontv/libfv.py

```diff
@@ -0,0 +1,384 @@
+    maintains version string state in memory, parses semicolon delimited substrings of the version string, and
+    provides public methods for version string modification and writes back to the font file on disk.
+
+    The class works on Python "strings".  String types indicated below refer to the Python2 unicode type and Python3
+    string type.
+
+    PUBLIC ATTRIBUTES:
+
+    contains_metadata: (boolean) boolean for presence of metadata in version string
+
+    contains_status: (boolean) boolean for presence of development/release status substring in the version string
+
+    develop_string: (string) The string to use for development builds in the absence of git commit SHA1 string
+
+    fontpath: (string) The path to the font file
+
+    is_development: (boolean) boolean for presence of development status substring at version_string_parts[1]
+
+    is_release: (boolean) boolean for presence of release status status substring at version_string_parts[1]
+
+    metadata: (list) A list of metadata substrings in the version string. Either version_string_parts[1:] or empty list
+
+    release_string: (string) The string to use for release builds in the absence of git commit SHA1 string
+
+    sha1_develop: (string) The string to append to the git SHA1 hash string for development builds
+
+    sha1_release: (string) The string to append to the git SHA1 hash string for release builds
+
+    status: (string) The development/release status string
+
+    ttf: (fontTools.ttLib.TTFont) for font file on fontpath file path
+
+    version_string_parts: (list) List that maintains in memory semicolon parsed substrings of font version string
+
+    version: (string) The version number substring formatted as "Version X.XXX"
+
+
+    PRIVATE ATTRIBUTES
+
+    _nameID_5_dict: (dictionary) {(platformID, platEncID,langID) : fontTools.ttLib.TTFont name record ID 5 object } map
+
+    :parameter fontpath: (string) file path to the .otf or .ttf font file
+
+    :parameter develop: (string) the string to use for development builds in the absence of git commit SHA1 string
+
+    :parameter release: (string) the string to use for release builds in the absence of a git commit SHA1 string
+
+    :parameter sha1_develop: (string) the string to append to the git SHA1 hash string for development builds
+
+    :parameter sha1_release: (string) the string to append to the git SHA1 hash string for release builds
+
+    :raises: fontTools.ttLib.TTLibError if fontpath is not a ttf or otf font
+
+    :raises: IOError if fontpath does not exist
+    """
+    def __init__(self, fontpath, develop="DEV", release="RELEASE", sha1_develop="-dev", sha1_release="-release"):
+        self.fontpath = fontpath
+        self.develop_string = develop
+        self.release_string = release
+        self.sha1_develop = sha1_develop
+        self.sha1_release = sha1_release
+
+        self.ttf = ttLib.TTFont(self.fontpath)
```

Comment:
It would be good if we could `__init__` this by passing an instance of `ttLib.TTFont` instead of `fontpath`. Because I think @felipesanches is planning to use `FontVersion` in Font Bakery, and there we have already `TTFont` instances in memory. (@felipesanches correct me if I'm wrong with my thinking.)

Reply:
How do you think it would be best to include both options (file path and ttLib.TTFont) as part of the instantiation parameters?

```
def __init__(self, fontpath=None, ttfont=None, develop="DEV", release="RELEASE", sha1_develop="-dev", sha1_release="-release"):
```

and then have it explicitly defined by calling code, fail if both remain `None`?

Reply:
I think I remember there was an example somewhere in the wild. I'll try to dig it up now …

Reply:
Thanks Lasse!

Reply:
This should be a simple if/else block around the current assignment of the ttLib.TTFont during instantiation so not a big deal if this is helpful.  This takes place before the remainder of the code to pull the version string in so everything else should stay the same.

Reply:
Having `fontpath` and `ttfont` explicitly is fine for me. 

However, I found what I was reminded of, it's not the same but in the `TTFont` constructor itself, LOL:

https://github.com/fonttools/fonttools/blob/master/Lib/fontTools/ttLib/__init__.py#L65-L72

```
		The constructor can be called with a few different arguments.
		When reading a font from disk, 'file' should be either a pathname
		pointing to a file, or a readable file object
```

Thus `FontVersion` could additionally allow a `file object` as argument.

I would try to go with just one argument and fill the rest with what is there.  But it depends on the semantics of `self.fontpath` (what else is it used for, if not just opening the file, sorry, I didn't read it all to give good advice).

Assuming `fontpath` is mandatory (is it?):
```py
def __init__(self, font, ...):
    try:
        self.fontpath = font.reader.file.name
        self.ttf = font
    except AttributeError: 
        # doesn't look like a TTFont instance with a file
        # expecting a string or a file
        self.ttf = ttLib.TTFont(font)
        # this fails with an AttributeError if font was a file like object but without
        # a name. I'm not sure how important self.fontpath is for the operation
        # of FontVersion
        self.fontpath = self.ttf.reader.file.name
```

Reply:
`fontpath` is only mandatory in the sense that we need to instantiate a `ttLib.TTFont` object and to have the original path to write out changes in the font version string to the original font binary.

I've never used `ttLib.TTFont.reader.file.name`.  This returns the filepath to the font file that was used to instantiate the `TTFont` object?

Reply:
Just confirmed that this is the case, it does maintain the relative path to the font binary.

Reply:
added support for this in https://github.com/source-foundry/font-v/pull/15/commits/e99666012b5e2cb52086aca29680bef32491661b

That works really well Lasse.  Thank you very much!  Will add tests for the new instantiation approach tonight.

Reply:
`reader.file` is like the return value of `file = open(file, "rb")` if opened in `TTFile.__init__` i.e.:

```py
>>> file = open('../fonts/apache/roboto/Roboto-Regular.ttf', 'rb')
>>> file.name
'../fonts/apache/roboto/Roboto-Regular.ttf'
```

But an edge case here is that a file-like object can not have a name attribute, e.g. a [io.BytesIO](https://docs.python.org/2/library/io.html#io.BytesIO):

```py
>>> file1 = open('../fonts/apache/roboto/Roboto-Regular.ttf', 'rb')
>>> file1.name
'../fonts/apache/roboto/Roboto-Regular.ttf'
>>> from io import BytesIO
>>> file2 = BytesIO(file1.read())
# no name!
>>> file2.name
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: '_io.BytesIO' object has no attribute 'name'

>>> from fontTools.ttLib import TTFont

>>> file1.seek(0) # rewind
>>> font1 = TTFont(file)
>>> font1.reader.file.name
'../fonts/apache/roboto/Roboto-Regular.ttf'

# still no name!
>>> font2 = TTFont(file2)
>>> font2.reader.file.name
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: '_io.BytesIO' object has no attribute 'name'

# but it is a good ttFont:
>>> font2['name'].names[0].toUnicode()
u'Copyright 2011 Google Inc. All Rights Reserved.'
```
```

Reply:
Interesting.  Something that would be common enough in any use case to warrant an attempt to handle this?  That seems like an awful lot of work to instantiate the TTFont object when it is as simple as passing a file path string or a file handle.

Reply:
I don't think it needs handling now, it can still be fixed when a real use case comes up.

In the Font Bakery Dashboard, I'd love to have all of Font Bakery running without disk I/O (because the font files are not on the Font Bakery-Worker disks initially, just in memory). But, there are more than enough dependencies that expect reading font files from disk, it won't be any soon that we can run without disk i/o, if ever.

Reply:
Thanks for the suggestion Lasse.  It's now supported and tests are in place for this.  Will part of the initial release of this library.  I really appreciate this feedback.  This was helpful.